src/MessageTranslator.cpp

```diff
@@ -0,0 +1,281 @@
+    std::vector<uint8_t> payloadData;
+    for(size_t iCount = 0 ; iCount < uMessage->payload().size(); iCount++) {
+        payloadData.push_back(uMessage->payload().data()[iCount]);
+    }
+    payload->set_data(payloadData);
+    someIpMessage->set_payload(payload);
+
+    return someIpMessage;
+}
+
+/**
+  * @brief @return @see @ref MessageTranslator::buildUPayload()
+  * @param[in] someIpMessage @see @ref MessageTranslator::buildUPayload()
+  */
+std::shared_ptr<UPayload> MessageTranslator::buildUPayload(std::shared_ptr<message> const someIpMessage) const {
+
+    const std::shared_ptr<payload> someipPayload = someIpMessage->get_payload();
+    std::stringstream its_message;
+
+    for (uint32_t i = 0U; i < someipPayload->get_length(); ++i) {
+        its_message << static_cast<int>(someipPayload->get_data()[i]);
+    }
+
+    std::shared_ptr<UPayload> uPayload =
+        std::make_shared<UPayload>(
+            reinterpret_cast<uint8_t const *>(its_message.str().c_str()),
+            its_message.str().size() + 1, UPayloadType::VALUE);
+
+    return uPayload;
+ }
+
+/**
+  * @brief @return @see @ref MessageTranslator::buildUAttribute()
+  * @param[in] uriSource @see @ref MessageTranslator::buildUAttribute()
+  * @param[in] uriSink @see @ref MessageTranslator::buildUAttribute()
+  * @param[in] generatedUUID @see @ref MessageTranslator::buildUAttribute()
+  * @param[in] messageType @see @ref MessageTranslator::buildUAttribute()
+  */
+std::shared_ptr<UAttributes> MessageTranslator::buildUAttribute(
+    UUri const &uriSource,
+    uprotocol::v1::UUri const &uriSink,
+    UUID const &generatedUUID,
+    message_type_e const &messageType,
+    UPriority priority) const {
+
+    //TODO: Pass correct URI according to 1_5_7 to UAttributesBuilder
+    UMessageType const type = convertSomeipToUMessageType(messageType);
+    UAttributesBuilder builder(uriSource ,generatedUUID, type, priority);
+    builder.setSink(uriSink);
+    std::shared_ptr<UAttributes> attributes = std::make_shared<UAttributes>(builder.build());
+    return attributes;
+}
+
+/**
+  * @brief @return @see @ref MessageTranslator::translateSomeipToUMsgForNotification()
+  * @param[in] someIpMessage @see @ref MessageTranslator::translateSomeipToUMsgForNotification().
+  * @param[in] strTopic @see @ref MessageTranslator::translateSomeipToUMsgForNotification().
+*/
+std::shared_ptr<UMessage> MessageTranslator::translateSomeipToUMsgForNotification(
+    std::shared_ptr<message> const someIpMessage,
+    UEntity const &uEntity ,
+    UAuthority const &uAuthority,
+    UResource const &uResource) const {
+
+    LogInfo("{} serviceid[0x{:x}] instance[0x{:x}] eventgrpid[0x{:x}]",
+            __FUNCTION__, someIpMessage->get_service(), someIpMessage->get_instance(), someIpMessage->get_method());
+
+    //build URI
+    std::shared_ptr<UUri> uriPtr = std::make_shared<UUri>();
+    uriPtr->mutable_authority()->CopyFrom(uAuthority);
```

Comment:
There are two ways that protobuf allows to set values
-set_allocated_X
-mutable_X

The set_allocated_authority API in the proto only accepts a raw pointer (e.g., UAuthority*), but we have not used raw pointers anywhere in our project. It is not recommended to use raw pointers to avoid memory leakage, especially in C++ standard 11 and above. Instead, the function parameter reference (UAuthority const &uAuthority) points to a smart pointer, such as std::shared_ptr<UAuthority>, and cannot be directly passed to the set_allocated_authority() API which only accepts raw pointer.
To set the authority to a URI, the recommended approach is to use mutable_authority() to obtain a mutable reference to the authority field in the protobuf message. Then, you can copy the contents using the CopyFrom API. This ensures proper memory management and avoids memory leaks.
 
By using mutable_authority() and CopyFrom, you can safely set the authority field in the protobuf message without relying on raw pointers and avoiding memory leakage.